app/code/community/Bolt/Boltpay/Helper/Coupon.php

```diff
@@ -0,0 +1,662 @@
+
+            // Try applying to parent first
+            $this->applyCouponToQuote($parentQuote, $couponCode);
+            // Apply coupon to clone
+            $this->applyCouponToQuote($immutableQuote, $couponCode);
+        } catch (\Exception $e) {
+            $this->setErrorResponseAndThrowException(
+                self::ERR_SERVICE,
+                $e->getMessage(),
+                422,
+                @$immutableQuote
+            );
+        }
+    }
+
+    /**
+     * Applies the coupon code to the quote and recalculates totals.
+     *
+     * @param Mage_Sales_Model_Quote $quote
+     * @param string $couponCode
+     */
+    protected function applyCouponToQuote($quote, $couponCode)
+    {
+        $quote->getShippingAddress()->setCollectShippingRates(true);
+        $quote->setCouponCode($couponCode);
+
+        Mage::helper('boltpay')->collectTotals($quote, true)->save();
+    }
+
+    /**
+     * Verifies that the coupon code was applied to the quote.
+     *
+     * @throws Exception     thrown if coupon in quote doesn't match the coupon applied.
+     */
+    protected function validateAfterApplyingCoupon()
+    {
+        $immutableQuote = $this->getImmutableQuote();
+        if ($immutableQuote->getCouponCode() != $this->getCouponCode()) {
+            $this->setErrorResponseAndThrowException(
+                self::ERR_SERVICE,
+                sprintf("Invalid coupon code response for coupon [%s]", $this->getCouponCode()),
+                422,
+                $immutableQuote
+            );
+        }
+    }
+
+    /**
+     * Sets the response error and the http status code and then throws an exception.
+     * @param int $errCode
+     * @param string $message
+     * @param int $httpStatusCode
+     * @throws Exception
+     */
+    protected function setErrorResponseAndThrowException($errCode, $message, $httpStatusCode)
+    {
+        $this->responseError = [
+            'code' => $errCode,
+            'message' => $message,
+        ];
+
+        $this->httpCode = $httpStatusCode;
+        $this->responseCart = $this->getCartTotals();
+        throw new \Exception($message);
+    }
+
+    /**
+     * Gets all cart totals from Bolt buildCart.
+     *
+     * @return array
+     */
+    protected function getCartTotals()
+    {
+        $quote = $this->getImmutableQuote();
+        /** @var Bolt_Boltpay_Helper_Api $boltHelper */
+        $boltHelper = Mage::helper('boltpay/api');
+        $items = @$quote->getAllVisibleItems();
+
+        $cart = $boltHelper->buildCart($quote, $items, 'multi-page');
+        return [
+            'total_amount' => $cart['total_amount'],
+            'tax_amount' => $cart['tax_amount'],
+            'discounts' => $cart['discounts'],
+        ];
+    }
+
+    /**
+     * Sets the success response (discount code, amount, description, and type). Also sets the HTTP status to 200.
+     *
+     * @return array
+     */
+    protected function setSuccessResponse()
```

Comment:
Please make sure to also include common response header - https://github.com/BoltApp/bolt-magento1/blob/effeb4e1b91c5195438f0b4ddf77f9281e51a8ee/app/code/community/Bolt/Boltpay/Helper/Api.php#L1063